rueidisprob/slidingbloomfilter.go

```diff
@@ -0,0 +1,443 @@
+
+	for _, key := range keys {
+		h1, h2 := hash([]byte(key))
+		for i := uint(0); i < s.hashIterations; i++ {
+			offset := len(*buf)
+			*buf = strconv.AppendUint(*buf, index(h1, h2, i, size), 10)
+			allIndexes = append(allIndexes, rueidis.BinaryString((*buf)[offset:]))
+		}
+	}
+	return allIndexes
+}
+
+func (s *slidingBloomFilter) Exists(ctx context.Context, key string) (bool, error) {
+	exists, err := s.ExistsMulti(ctx, []string{key})
+	if err != nil {
+		return false, err
+	}
+
+	return exists[0], nil
+}
+
+func (s *slidingBloomFilter) ExistsMulti(ctx context.Context, keys []string) ([]bool, error) {
+	if len(keys) == 0 {
+		return nil, nil
+	}
+
+	err := s.checkAndRotate(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check rotation: %w", err)
+	}
+
+	buf := bytesPool.Get(0, len(keys)*int(s.hashIterations)*8)
+	defer bytesPool.Put(buf)
+
+	indexes := s.indexes(keys, &buf.s)
+
+	args := make([]string, 0, len(indexes)+1)
+	args = append(args, s.hashIterationString)
+	args = append(args, indexes...)
+
+	resp := s.existsMultiScript.Exec(ctx, s.client, s.existsMultiKeys, args)
+	if resp.Error() != nil {
+		return nil, resp.Error()
+	}
+
+	arr, err := resp.ToArray()
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]bool, len(keys))
+	for i, el := range arr {
+		v, err := el.AsBool()
+		if err != nil {
+			if rueidis.IsRedisNil(err) {
+				result[i] = false
+				continue
+			}
+
+			return nil, err
+		}
+
+		result[i] = v
+	}
+	return result, nil
+}
+
+func (s *slidingBloomFilter) Reset(ctx context.Context) error {
+	resp := s.client.Do(ctx,
+		s.client.B().
+			Eval().
+			Script(slidingBloomFilterResetScript).
+			Numkeys(4).
+			Key(s.addMultiKeys...).
+			Build(),
+	)
+	return resp.Error()
+}
+
+func (s *slidingBloomFilter) Delete(ctx context.Context) error {
+	resp := s.client.Do(ctx, s.client.B().Del().Key(s.rotateKeys...).Build())
+	return resp.Error()
+}
+
+func (s *slidingBloomFilter) Count(ctx context.Context) (uint64, error) {
+	resp := s.client.Do(
+		ctx,
+		s.client.B().
+			Get().
+			Key(s.counter).
+			Build(),
+	)
+	count, err := resp.AsUint64()
+	if err != nil {
+		if rueidis.IsRedisNil(err) {
+			return 0, nil
+		}
+
+		return 0, err
+	}
+
+	return count, nil
+}
+
+func (s *slidingBloomFilter) RotationErrors() <-chan error {
+	s.rotationErrorsOnce.Do(func() {
+		s.rotationErrors = make(chan error, 1)
+	})
+	return s.rotationErrors
+}
+
+func (s *slidingBloomFilter) checkAndRotate(ctx context.Context) error {
+	// Fast path - check if enough time has passed without locking
+	if time.Since(s.lastRotationAttempt) < s.rotation {
+		return nil
+	}
+
+	// Slow path - double check with lock
+	s.rotationMu.Lock()
+	now := time.Now()
+	if now.Sub(s.lastRotationAttempt) < s.rotation {
+		s.rotationMu.Unlock()
+		return nil
+	}
+	s.lastRotationAttempt = now
+	s.rotationMu.Unlock()
+
+	s.attempts = 0
+	for {
+		resp := s.rotateScript.Exec(ctx, s.client, s.rotateKeys, s.rotateArgs)
+		if resp.Error() == nil {
```

Comment:
doesn't we need to check rotation is success?

Reply:
To avoid having to block until the rotation is done, we set the timer even before the rotation is done so that other calls will return, leaving only one that will retry the rotation until it finishes or reaches max, skipping it to the window and that could be a problem... maybe an additional goroutine that maintains the rotation would be better?, WDYT @proost @rueian ?

Reply:
Extreme edge case:
If `rotation` is short(e.g. 500ms), there are multiple goroutines and timeout in context bit long and executing script might long,

1. race condition happen in `attempts`.
2. no error but "0" returned.

I think:
1. appliying single flight
2. make L428 to L441 to critical section
Or something other choice are ok.

Reply:
As you said, leaving only one is critical.

Reply:
Let me clarify this:

rotation is important. but i think trying rotation less in each application server can be applied to here even extreme case.

I think lazy rotation is enough.

Reply:
It's not a problem if multiples goroutines try to rotate in Redis (but it should not happen) the lock will avoid rotating several times in those cases. We could track the last rotation and set it back in case of failures, meaning that the next call will try to rotate again.

The main idea is to don't block on the rotation path several calls, they are safe to go to Redis

Reply:
OK, I understand your intention. Then it skip it.